### PR TITLE
Feature/user profile serialization and repository

### PR DESCRIPTION
**Pull request by unusedusername01**
This pull request introduces a comprehensive suite of Android instrumented tests for the `UserProfileRepositoryFirestore` class, adds robust Firebase emulator utilities, and extends the `UserProfile` data model with serialization and photo handling capabilities. It also updates the repository interface to support searching user profiles and configures the app for emulator networking. These changes significantly improve the testability, reliability, and flexibility of user profile management in the app.

**Testing infrastructure and utilities:**

* Added `UserProfileRepositoryFirestoreTest.kt` with extensive tests covering user profile creation, updating, deletion, access control, and search functionality, including edge cases and constraints.
* Introduced `FirebaseEmulator.kt` utility to manage Firebase Auth and Firestore emulators for Android tests, including connection detection, test user sign-in, and clearing emulator state.

**Model and interface enhancements:**

* Extended `UserProfile.kt` to support nullable fields, serialization/deserialization to Firestore, and efficient Bitmap-to-Blob conversion for profile photos with size constraints.
* Updated `UserProfileRepository` interface to include a `searchUserProfiles` method, returning results as a `Flow` for real-time updates and supporting query/section filtering. [[1]](diffhunk://#diff-cd1dce209ab0742a209c7589d6bb36858683e45066305c4f3eafd870a4cf0fcbR3-R4) [[2]](diffhunk://#diff-cd1dce209ab0742a209c7589d6bb36858683e45066305c4f3eafd870a4cf0fcbR48-R63)

**Configuration:**

* Enabled cleartext traffic in `AndroidManifest.xml` to support emulator networking.

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/model/profile/UserProfileRepositoryFirestoreTest.kt [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+  @Test
+  fun cannotDeleteOtherUsersProfile() = runTest {
+    try {
+      repository.deleteUserProfile("some-other-user-id")
+      fail("Expected IllegalArgumentException when deleting another user's profile")
+    } catch (e: IllegalArgumentException) {
+      // Expected exception
+    }
+  }
+
```
Comment (Funoma) [inline app/src/androidTest/java/com/android/sample/model/profile/UserProfileRepositoryFirestoreTest.kt:241]: You have already written tests for the person who will implement this feature in the future, that’s really thoughtful ! 👍

**Review by Funoma — COMMENTED** [review]
Really good PR! There are just a few minor implementation details to review in the future, but overall it looks great.

File: app/src/main/java/com/android/sample/model/profile/UserProfile.kt [hunk 2]
```diff
@@ -1,18 +1,130 @@
...
+ *   favorites etc.)
+ */
 data class UserProfile(
     val id: String,
     val name: String,
     val lastName: String,
-    val email: String,
-    val photo: Bitmap,
+    val email:
+        String?, // User (or default settings) can choose to not share email with others -> nullable
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfile.kt:43]: Nice catch, I'll try to add this in my next PR :)

File: app/src/main/java/com/android/sample/model/profile/UserProfile.kt [hunk 3]
```diff
@@ -1,18 +1,130 @@
...
+    val photo: Bitmap?, // Nullable Bitmap for user photo in case user hasn't set one
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfile.kt:44]: I was considering that as well. At this stage, I’m not sure whether it’s better to fetch a default image from local storage within the model or to handle it directly in the View when the value is null. I’ll bring this up in our next stand-up with the design team. Thanks for the suggestion!

File: app/src/main/java/com/android/sample/model/profile/UserProfileRepository.kt [hunk 4]
```diff
@@ -43,4 +45,20 @@ interface UserProfileRepository {
...
+   *
+   * @param query The search query string.
+   * @param section Optional section to filter the search results.
+   * @param resultsPerPage The maximum number of results to return per page (default is 20).
+   * @return A Flow emitting UserProfile objects that match the search criteria.
+   */
+  suspend fun searchUserProfiles(
+      query: String,
+      section: Section?,
+      resultsPerPage: Int = 20,
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfileRepository.kt:62]: Do you mean outside the function's scope ? (e.g: As a class constant ?)

File: app/src/main/java/com/android/sample/model/profile/UserProfileRepository.kt [hunk 5]
```diff
@@ -43,4 +45,20 @@ interface UserProfileRepository {
...
+  ): Flow<UserProfile>
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfileRepository.kt:63]: My bad on this one, this likely a typo. The intention is to eventually have a single page that dynamically loads a set number of items in one composable, such as a `LazyColumn`. I’ve renamed the argument to `resultsPerView` to better reflect this purpose.

File: app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  // Path structure: "public_profiles/{userId}" and "private_profiles/{userId}" (userId same in
+  // both)
+
+  private val publicCollectionRef = db.collection(PUBLIC_PROFILES_PATH)
+  private val privateCollectionRef = db.collection(PRIVATE_PROFILES_PATH)
+
+  // Near zero collision probability for user profile IDs
+  override fun getNewUid(): String {
+    return UUID.randomUUID().toString()
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt:33]: Noted 👍

File: app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  // Retrieves all public user profiles
+  override suspend fun getAllUserProfiles(): List<UserProfile> {
```
[the remark this answers is not in the document]
Comment (unusedusername01) [reply app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt:37]: I guess you're right, we'll work on a filter once the Model structure is refined.
